### DISCUSSION TITLE
Implement some metrics through System.Diagnostics.DiagnosticSource

### DISCUSSION THREAD

**Discussion by aygalinc** [rabbitmq/rabbitmq-dotnet-client#1730] [General]
### Is your feature request related to a problem? Please describe.

Actually there is no metric provided by rabbitmq client and observability must done done on application side if I want to have have. It would benefitiate all the user if rabbitMQ lib can expose some sort of metrics

### Describe the solution you'd like

Add a dedicated metricreporter that rely on standard System.Diagnostics.DiagnosticSource that can expose various metrics related to publishing or consuming.
It can be easily integrated by user of OTEL or Prom without additionnal library

**Comment by michaelklishin** [comment]
@aygalinc you are welcome to contribute such a "dedicated metric reporter". OpenTelemetry support was contributed by multiple community members. 

One obvious first step is deciding what metrics would be the most important to expose and how that would correlate to the OpenTelemetry-related feature set.

Right now there aren't any specifics to work with so according to the standards of our team, this is not an actionable issue but rather a discussion of some ideas.

**Reply by aygalinc** [reply]
I duplicate this in an issue, i will try to submit a PR when I got some slack.

**Reply by BOLT04** [reply]
Hi everyone, I was searching for the metrics support of this library, since it's something my team would like and stumbled on this discussion 🙂.
 
> One obvious first step is deciding what metrics would be the most important to expose and how that would correlate to the OpenTelemetry-related feature set.

We could start looking at this PR rabbitmq/rabbitmq-amqp-dotnet-client#84 and implement the same metrics for `RabbitMQ.Client`. It was done by @aygalinc and seems like an amazing starting point. In addition to the metrics of that PR, I'd like to suggest these too (Java SDK has some at least):
- Number of ACK messages (only used for publisher confirms)
- Number of NACK messages (only used for publisher confirms)
- Number of Connection and Channel shutdowns
- Number of messages sent to dead letter

Then we could take a look at these conventions from OpenTelemetry:
- [Messaging semantic conventions](https://opentelemetry.io/docs/specs/semconv/messaging/messaging-metrics/)
- [Semantic Conventions for RabbitMQ](https://opentelemetry.io/docs/specs/semconv/messaging/rabbitmq/) 

They are both experimental so they can change easily, the latest version is `1.30.0` at the time of writing. The `MetricsReporter` class on that PR seems to be inspired by conventions of an earlier version, but also has quite some custom metrics. I'm unfamiliar with the current roadmap on OpenTelemetry regarding these conventions, or who is working on them. So we could decide to follow along with the metrics instrumented in other RabbitMQ SDKs while discussing in the background some standard to follow.

@aygalinc did you start working on a draft? I'm happy to help when I get some free time too.

**Reply by aygalinc** [reply]
Hi ! I just check the code base and I think the most straightforward is to design a wrapper around the event counter. ( Or replace it directly with a metric reporter because event counter seems deprecated in recent DotNet) 

I basically had no time to do this but maybe when I get some bandwidth on day 😊
